### PR TITLE
MongoDB integration

### PR DESCRIPTION
**Pull request by jochman**
## Status
- [x] Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/20026

## Minimum version of Demisto
- [x] 4.5.0
## Does it break backward compatibility?
   - [x] No

## Must have
- [x] Tests
- [x] Documentation 

Documentation:
https://github.com/demisto/content/blob/mongodb-integration/Integrations/MongoDB/README.md

### CODE REVIEW

**Comment by jochman** [comment]
@kirbles19 
https://github.com/demisto/content/blob/mongodb-integration/Integrations/MongoDB/MongoDB.yml

File: Integrations/MongoDB/CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1 @@
+## [Unreleased]
```
Comment (Itay4) [inline Integrations/MongoDB/CHANGELOG.md]: no need in changelog for new integration

File: Integrations/MongoDB/MongoDB.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+        self._client = MongoClient(urls, username=username, password=password, ssl=ssl)
+        self.db: Database = self._client.get_database(database)
+
+    def is_collection_in_db(self, collection: str) -> bool:
+        return collection in self.db.list_collection_names()
+
+    def get_collection(self, collection: str) -> Collection:
+        if self.is_collection_in_db(collection):
+            return self.db.get_collection(collection)
+        return_error(
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.py]: raise exception instead

File: Integrations/MongoDB/MongoDB.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+        "mongodb-get-entry-by-id": get_entry_by_id_command,
+        "mongodb-query": search_query,
+        "mongodb-insert": insert_entry_command,
+        "mongodb-update": update_entry_command,
+        "mongodb-delete": delete_entry_command,
+        "mongodb-list-collections": list_collections_command,
+        "mongodb-create-collection": create_collection_command,
+        "mongodb-drop-collection": drop_collection_command,
+    }
+    # The command demisto.command() holds the command sent from the user.
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.py]: remove comment

File: Integrations/MongoDB/MongoDB.yml [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+- display: Username
+  name: credentials
+  required: true
+  type: 9
+- additionalinfo: A comma separated list of urls with port.
+  display: Server URLs with port (host1.com:27017,host2.com:27017)
+  name: urls
+  required: true
+  type: 0
+- display: Database
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.yml]: database name?

File: Integrations/MongoDB/MongoDB.yml [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+      isArray: true
+      name: object_id
+      required: true
+      secret: false
+    deprecated: false
+    description: Get an entry from database by ID
+    execution: false
+    name: mongodb-get-entry-by-id
+    outputs:
+    - contextPath: MongoDB._id
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.yml]:
```suggestion
    - contextPath: MongoDB.ID
```

File: Integrations/MongoDB/MongoDB.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+            return self.db.get_collection(collection)
+        return_error(
+            f"MongoDB: Collection '{collection}' has not found in database '{self.db.name}'"
+        )
+
+    def get_entry_by_id(self, collection: str, alert_ids: List[str]) -> List[dict]:
+        collection_obj = self.get_collection(collection)
+        entries: List[dict] = list()
+        for alert_id in alert_ids:
+            results = collection_obj.find({"_id": ObjectId(alert_id)})
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.py]: why not use `find_one()` ?

File: Integrations/MongoDB/MongoDB.yml [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+  commands:
+  - arguments:
+    - default: false
+      description: Name of the collection do get entry from.
+      isArray: false
+      name: collection
+      required: true
+      secret: false
+    - default: false
+      description: An ObjectID to get.
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.yml]: elaborate how this can be retrieved

File: Integrations/MongoDB/MongoDB.yml [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+      type: String
+  - arguments:
+    - default: false
+      description: Name of the collection do query from.
+      isArray: false
+      name: collection
+      required: true
+      secret: false
+    - default: false
+      description: A JSON query to search in collection.
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.yml]: add example

File: Integrations/MongoDB/MongoDB.py [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+        argToList(params.get("urls")),
+        params.get("credentials", {}).get("identifier"),
+        params.get("credentials", {}).get("password"),
+        params["database"],
+        not demisto.params().get("insecure", False),
+    )
+    commands = {
+        "test-module": test_module,
+        "mongodb-get-entry-by-id": get_entry_by_id_command,
+        "mongodb-query": search_query,
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.py]: stay consistent with the `_command` suffix

File: Integrations/MongoDB/MongoDB.py [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+        raw_response = client.query(collection, query_json)
+        readable_outputs = tableToMarkdown(
+            f"Total of {len(raw_response)} found in MongoDB collection `{collection}` with query: {query}:",
+            [entry.get("_id") for entry in raw_response],
+            headers=["_id"],
+        )
+        outputs = {CONTEXT_KEY: raw_response}
+        return readable_outputs, outputs, raw_response
+    except JSONDecodeError:
+        return_error("MongoDB: the `query` parameter is not a valid json.")
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.py]: raise exception instead

File: Integrations/MongoDB/MongoDB.py [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+        raw_response,
+    )
+    outputs = {CONTEXT_KEY: raw_response}
+    return readable_outputs, outputs, raw_response
+
+
+def search_query(
+    client: Client, collection: str, query: str, **kwargs
+) -> Tuple[str, dict, list]:
+    # test if query is a valid json
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.py]: whats with the comment?

File: Integrations/MongoDB/MongoDB.yml [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+      isArray: false
+      name: query
+      required: true
+      secret: false
+    deprecated: false
+    description: Searches items by query
+    execution: false
+    name: mongodb-query
+    outputs:
+    - contextPath: MongoDB._id
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.yml]:
```suggestion
    - contextPath: MongoDB.ID
```

File: Integrations/MongoDB/MongoDB.yml [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+      description: A JSON query to search in collection.
+      isArray: false
+      name: query
+      required: true
+      secret: false
+    deprecated: false
+    description: Searches items by query
+    execution: false
+    name: mongodb-query
+    outputs:
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.yml]: are these all the outputs? just ids?

File: Integrations/MongoDB/MongoDB.yml [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+      isArray: false
+      name: entry
+      required: true
+      secret: false
+    deprecated: false
+    description: Inserts an entry to the database
+    execution: false
+    name: mongodb-insert
+    outputs:
+    - contextPath: MongoDB._id
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.yml]:
```suggestion
    - contextPath: MongoDB.ID
```

File: Integrations/MongoDB/MongoDB.py [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+            object_ids = convert_object_id_to_str(results.inserted_ids)
+            human_readable = tableToMarkdown(
+                f"MongoDB: Successfully entered {len(object_ids)} entry to the '{collection}' collection.",
+                object_ids,
+                headers=["_id"],
+            )
+            outputs = {CONTEXT_KEY: [{"_id": _id} for _id in object_ids]}
+            return human_readable, outputs, object_ids
+    except JSONDecodeError:
+        return_error("MongoDB: The `entry` parameter is not a valid json.")
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.py]: raise exception instead

File: Integrations/MongoDB/MongoDB.yml [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+      type: String
+  - arguments:
+    - default: false
+      description: Name of the collection to insert entry from.
+      isArray: false
+      name: collection
+      required: true
+      secret: false
+    - default: false
+      description: Entry JSON formatted. can include `_id` argument or not.
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.yml]: add example

File: Integrations/MongoDB/MongoDB.py [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+) -> Tuple[str, dict, list]:
+    # test if query is a valid json
+    try:
+        entry_json = json.loads(entry)
+        if not isinstance(entry_json, list):
+            entry_json = [entry_json]
+        entries = convert_id_to_object_id(entry_json)
+        results = client.insert_entry(collection, entries)  # type: ignore[arg-type]
+        if not results.acknowledged:
+            return_error("MongoDB: Error occurred when trying to enter insert entries.")
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.py]: raise exception

File: Integrations/MongoDB/MongoDB.py [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+
+
+def delete_entry_command(
+    client: Client, collection, filter, delete_one, **kwargs
+) -> Tuple[str, None]:
+    try:
+        json_filter = json.loads(filter)
+        json_filter = convert_id_to_object_id(json_filter)
+    except JSONDecodeError:
+        return_error("MongoDB: The `filter` parameter is not a valid json.")
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.py]: EXCEPTION

File: Integrations/MongoDB/MongoDB.py [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+    results = client.delete_entry(collection, json_filter, argToBoolean(delete_one))
+    if not results.acknowledged:
+        return_error("MongoDB: Error occurred when trying to enter delete entries.")
+    return f"### MongoDB: Delete {results.deleted_count} entries.", None
+
+
+def create_collection_command(client: Client, collection: str, **kwargs) -> Tuple[str, None]:
+    if client.is_collection_in_db(collection):
+        return f"### MongoDB: Collection '{collection}' is already exists", None
+    _ = client.create_collection(collection)
```
Comment (Itay4) [inline Integrations/MongoDB/MongoDB.py]: why are we initalizing here `_`? just dont return anything

File: Integrations/MongoDB/README.md [hunk 20]
```diff
@@ -0,0 +1,336 @@
+## Overview
+---
+
+Use MongoDB to search and query entries
+This integration was integrated and tested with version xx of MongoDB
```
Comment (Itay4) [inline Integrations/MongoDB/README.md]: put in version here

File: TestPlaybooks/playbook-MongoDBTest.yml [hunk 21]
```diff
@@ -0,0 +1,1190 @@
+elasticcommonfields: {}
+id: TestMongoDB
```
Comment (Itay4) [inline TestPlaybooks/playbook-MongoDBTest.yml]:
```suggestion
id: MongoDB - Test
```

File: TestPlaybooks/playbook-MongoDBTest.yml [hunk 22]
```diff
@@ -0,0 +1,1190 @@
...
+version: -1
+name: TestMongoDB
```
Comment (Itay4) [inline TestPlaybooks/playbook-MongoDBTest.yml]:
```suggestion
name: MongoDB - Test
```

File: Integrations/MongoDB/MongoDB.py (hunk 6 above)
Comment (jochman) [reply Integrations/MongoDB/MongoDB.py]: Because this command allows entering multiple ids.

File: Integrations/MongoDB/MongoDB.py (hunk 11 above)
Comment (jochman) [reply Integrations/MongoDB/MongoDB.py]: it's for the try/catch, clarification.

File: Integrations/MongoDB/MongoDB.py (hunk 10 above)
Comment (jochman) [reply Integrations/MongoDB/MongoDB.py]: but the return_error is already doing it. Why should I raise an exception instead?

File: Integrations/MongoDB/MongoDB.py (hunk 19 above)
Comment (jochman) [reply Integrations/MongoDB/MongoDB.py]: it's returning collection. I'm by assigning this to nothing the output of the function is not printed to the console.

File: Integrations/MongoDB/MongoDB.yml (hunk 5 above)
Comment (jochman) [reply Integrations/MongoDB/MongoDB.yml]: i wanted to stary consistent on how it looks in MongoDB.

File: Integrations/MongoDB/MongoDB.yml (hunk 13 above)
Comment (jochman) [reply Integrations/MongoDB/MongoDB.yml]: that's all the outputs that returned by default. anything else in this query will be returned as context.

File: Integrations/MongoDB/MongoDB.py (hunk 19 above)
Comment (Itay4) [reply Integrations/MongoDB/MongoDB.py]: so if you dont return anything, what does it do?

File: Integrations/MongoDB/MongoDB.yml (hunk 5 above)
Comment (Itay4) [reply Integrations/MongoDB/MongoDB.yml]: stay consistent with demisto

File: Integrations/MongoDB/MongoDB.py (hunk 6 above)
Comment (Itay4) [reply Integrations/MongoDB/MongoDB.py]: so the description is misleading

File: Integrations/MongoDB/MongoDB.yml (hunk 13 above)
Comment (Itay4) [reply Integrations/MongoDB/MongoDB.yml]: what is "anything else"?

File: Integrations/MongoDB/MongoDB.py (hunk 6 above)
Comment (jochman) [reply Integrations/MongoDB/MongoDB.py]: > looks good
> 
> 1. convert to pack
> 2. `"` -> `'`
> 3. no need for proxy param?
> 4. you seem to be populating not too much to context data (for example just IDs), populate more
> 5. use `return_error` only in the `except` clause. in the code raise exceptions
> 6. make sure you have docstrings and type hints to all functions
> 7. no unit test ???



> looks good
> 
> 1. convert to pack
> 2. `"` -> `'`
> 3. no need for proxy param?
> 4. you seem to be populating not too much to context data (for example just IDs), populate more
> 5. use `return_error` only in the `except` clause. in the code raise exceptions
> 6. make sure you have docstrings and type hints to all functions
> 7. no unit test ???

I prefer using `"` and stayed consistent (as we have not standard with it).
 There's no such thing as a proxy in MongoDB.
The data populated depends on the object returned. I can only assure `_id` will get back.
All of the missing docstring functions are just straight forward.
There' are not function here required unittest.

**Comment by kirbles19** [comment]
@jochman Done.

File: Packs/MongoDB/Integrations/MongoDB/MongoDB.py [hunk 23]
```diff
@@ -0,0 +1,305 @@
...
+        if self.is_collection_in_db(collection):
+            return self.db.get_collection(collection)
+        raise DemistoException(
+            f"Collection '{collection}' has not found in database '{self.db.name}'"
+        )
+
+    def get_entry_by_id(self, collection: str, alert_ids: List[str]) -> List[dict]:
+        collection_obj = self.get_collection(collection)
+        entries: List[dict] = list()
+        for alert_id in alert_ids:
```
Comment (Itay4) [inline Packs/MongoDB/Integrations/MongoDB/MongoDB.py]: handle empty response with `No results found`

File: Packs/MongoDB/Integrations/MongoDB/MongoDB.yml [hunk 24]
```diff
@@ -0,0 +1,195 @@
...
+    name: mongodb-update
+  - arguments:
+    - default: false
+      description: The name of the collection from which to delete an entry.
+      isArray: false
+      name: collection
+      required: true
+      secret: false
+    - default: false
+      description: 'A query that matches the document to delete, for example: `{"key": "value"}`.'
```
Comment (Itay4) [inline Packs/MongoDB/Integrations/MongoDB/MongoDB.yml]: add example of `_id`

File: Packs/MongoDB/Integrations/MongoDB/MongoDB.yml [hunk 25]
```diff
@@ -0,0 +1,195 @@
...
+    - default: false
+      description: 'The content with which to update the entry, in the format: {"$set": {"key": "value"}}. You can use Update Operators or Aggregation Pipeline. For more information, see the integration documentation.'
+      isArray: false
+      name: update
+      required: true
+      secret: false
+    - auto: PREDEFINED
+      default: false
+      defaultValue: 'true'
+      description: Whether to update a single entry. If false, will set all found entries. The default is true.
```
Comment (Itay4) [inline Packs/MongoDB/Integrations/MongoDB/MongoDB.yml]:
```suggestion
      description: Whether to update a single entry. If false, will set all found entries.
```

File: Packs/MongoDB/Integrations/MongoDB/MongoDB.yml (hunk 24 above)
Comment (jochman) [reply Packs/MongoDB/Integrations/MongoDB/MongoDB.yml]: done

File: Packs/MongoDB/Integrations/MongoDB/MongoDB.py (hunk 23 above)
Comment (jochman) [reply Packs/MongoDB/Integrations/MongoDB/MongoDB.py]: handled

File: Packs/MongoDB/Integrations/MongoDB/MongoDB.py [hunk 26]
```diff
@@ -0,0 +1,305 @@
...
+        if "_id" in entry:
+            entry["_id"] = ObjectId(entry["_id"])
+        return entry
+
+    if isinstance(entries, list):
+        return [_convert(entry) for entry in entries]
+    return _convert(entries)
+
+
+def convert_object_id_to_str(entries: List[ObjectId]) -> List[str]:
```
[the remark this answers is not in the document]
Comment (jochman) [reply Packs/MongoDB/Integrations/MongoDB/MongoDB.py]: done

**Comment by jochman** [comment]
> * push all `Entry` objects to be under `Collection` in the context
Can't do that with our DT, it'll replace all content of `Collection` instead of the entry itself.
I've pushed a `collection` key to each entry.

File: Integrations/MongoDB/CHANGELOG.md (hunk 1 above)
Comment (jochman) [reply Integrations/MongoDB/CHANGELOG.md]: The validator requires it

**Comment by Itay4** [comment]
1. handle empty results returned from `list-collections`: no results found in human readable and dont push `null` to context
2. add limit arg to query
3. check if theres timeout and set it
4. handle tls/ssl

**Comment by jochman** [comment]
moved https://github.com/demisto/content/pull/6266

**Pull request closed without merging**